### PR TITLE
Document various SQL drivers

### PR DESCRIPTION
**Pull request by cmouse**
DOV-4754

### CODE REVIEW

File: source/configuration_manual/sql.rst [hunk 1]
```diff
@@ -0,0 +1,526 @@
...
+
+   driver = <driver name>
+   connect_string = <driver specific options>
+
+in your dictionary or authentication configuration file.
+The connect_string is a space separated list of ``key=value`` pairs.
+
+Cassandra
+=========
+
```
Comment (sirainen) [inline source/configuration_manual/sql.rst:23]: I think these would be better split to their own pages. Now this Cassandra stuff takes almost the whole page.

File: source/configuration_manual/sql.rst [hunk 2]
```diff
@@ -0,0 +1,526 @@
...
+.. code::
+
+   driver = <driver name>
+   connect_string = <driver specific options>
+
+in your dictionary or authentication configuration file.
+The connect_string is a space separated list of ``key=value`` pairs.
+
+Cassandra
+=========
```
Comment (sirainen) [inline source/configuration_manual/sql.rst:22]: https://doc.dovecot.org/configuration_manual/mail_location/obox/dictmap_cassandra/ should link here.

File: source/configuration_manual/sql.rst [hunk 3]
```diff
@@ -0,0 +1,526 @@
...
+- Values: :ref:`string`
+
+Host or IP address to connect. Can appear multiple times.
+
+``port``
+--------
+- Default: ``9042``
+- Values: :ref:`uint`
+
+RPC port to use.
```
Comment (sirainen) [inline source/configuration_manual/sql.rst:45]: CQL native transport port to use.

File: source/configuration_manual/sql.rst [hunk 4]
```diff
@@ -0,0 +1,526 @@
...
+
+``ssl_verify``
+---------------
+- Default: ``none``
+- Values: ``none``, ``cert``, ``cert-ip``, ``cert-dns``
+
+Configure the peer certificate validation method.
+
+``none``
+  Disables validation
```
Comment (sirainen) [inline source/configuration_manual/sql.rst:110]: \* list here?

File: source/configuration_manual/sql.rst [hunk 5]
```diff
@@ -0,0 +1,526 @@
...
+- Values: ``critical``, ``error``, ``warn``, ``info``, ``debug``, ``trace``
+
+Driver log level.
+
+``debug_queries``
+-----------------
+- Default: no
+- Values: :ref:`boolean`
+
+Whether to log queries.
```
Comment (sirainen) [inline source/configuration_manual/sql.rst:133]: CQL queries

File: source/configuration_manual/sql.rst [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+-----------
+- Default: no
+- Values: :ref:`boolean`
+
+Whether to collect metrics.
+.. seealso:: :ref:`cassandra_metrics_json_output`
+
+``num_threads``
+----------------
+- Default: <driver dependant>
```
Comment (sirainen) [inline source/configuration_manual/sql.rst:146]: dependent

File: source/configuration_manual/sql.rst [hunk 7]
```diff
@@ -0,0 +1,526 @@
...
+``ssl_key``
+-----------
+- Default: <empty>
+- Values: :ref:`string`
+
+Path to private key matching ``ssl_cert`` to use for authenticating against the remote server.
+
+``ssl_cipher``
+---------------
+- Default: <library dependant>
```
Comment (sirainen) [inline source/configuration_manual/sql.rst:459]: dependent

**Review by sirainen — CHANGES_REQUESTED** [review]

File: source/configuration_manual/sql.rst [hunk 8]
```diff
@@ -0,0 +1,527 @@
...
+- Values: :ref:`uint`
+
+Set number of IO threads to handle query requests.
+
+``page_size``
+-------------
+- Default: -1
+- Values: :ref:`uint`
+
+Set the number of results per page.
```
Comment (sirainen) [inline source/configuration_manual/sql.rst:156]: `-1 = Disabled.` Would be useful also to describe the purpose of it .. but I'm not sure where to easily find a good clear explanation.

File: source/configuration_manual/sql.rst [hunk 9]
```diff
@@ -0,0 +1,527 @@
...
+
+Host to connect to.
+
+``hostaddr``
+------------
+Alias for ``host``.
+
+``port``
+--------
+- Default: ``9042``
```
Comment (sirainen) [inline source/configuration_manual/sql.rst:375]: 3306

File: source/configuration_manual/sql.rst [hunk 10]
```diff
@@ -0,0 +1,527 @@
...
+==========
+
+Driver name is ``postgresql``.
+
+To compile support for this driver, you need PostgreSQL client library and headers.
+
+Supported options
+-----------------
+
+PostgreSQL parses connection string directly. Doveot only parses ``host`` argument.
```
Comment (sirainen) [inline source/configuration_manual/sql.rst:495]: Dovecot

File: source/configuration_manual/sql.rst [hunk 11]
```diff
@@ -0,0 +1,527 @@
...
+
+To compile support for this driver. you need sqlite library and headers.
+
+Supported options
+-----------------
+
+.. versionchanged:: v2.3.18
+
+Prior v2.3.18, Dovecot uses the whole value as filename to connect, whitespace included.
+This was changed to support options, so it will only use initial non-whitespace part as filename.
```
Comment (sirainen) [inline source/configuration_manual/sql.rst:512]: Oh, didn't realize before, but there are no other settings in Dovecot that behave this way, better not add yet another different behavior. For example to make it work similarly as passwd-file you could just treat all parameters with "=" as options and when = isn't found that's the path. Or just that the last parameter is the path.

**Review by sirainen — CHANGES_REQUESTED** [review]

File: source/configuration_manual/sql/cassandra.rst [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+--------------
+- Default: ``warn``
+- Values: ``critical``, ``error``, ``warn``, ``info``, ``debug``, ``trace``
+
+Driver log level.
+
+``debug_queries``
+-----------------
+- Default: no
+- Values: :ref:`boolean`
```
Comment (sirainen) [inline source/configuration_manual/sql/cassandra.rst:112]: As mentioned in https://github.com/dovecot/documentation/pull/301 the existence of the setting is enough to enable this, so using "no" still enables this. Should we clarify?

File: source/configuration_manual/sql.rst (hunk 4 above)
Comment (cmouse) [reply source/configuration_manual/sql.rst]: this is a dd/dt list

Comment (cmouse) [reply source/configuration_manual/sql.rst]: "keyword definition list"

File: source/configuration_manual/sql/cassandra.rst [hunk 13]
```diff
@@ -0,0 +1,333 @@
...
+
+Whether to log CQL queries. Note that this setting behaves differently than other boolean settings.
+The feature is enabled by presence of ``debuq_queries`` in connect string, so
```
Comment (sirainen) [inline source/configuration_manual/sql/cassandra.rst:117]: debug

File: source/configuration_manual/sql/cassandra.rst [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+to disable this feature, you must remove the keyword completely.
+
+``metrics``
+-----------
+- Default: no
+- Values: :ref:`boolean`
```
Comment (sirainen) [inline source/configuration_manual/sql/cassandra.rst:123]: This is a string, pointing to a path where to write metrics.

File: source/configuration_manual/sql/cassandra.rst [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+Write consistency when deleting from the database fails with primary consistency.
+
+.. seealso:: :ref:`dictmap_cassandra_fallback_consistency`
+
+``latency_aware_routing``
+-------------------------
+- Default: no
+- Values: :ref:`boolean`
+
+When turned on, latency-aware routing tracks the latency of queries to avoid sending new queries to poorly performing Cassandra nodes.
```
Comment (sirainen) [inline source/configuration_manual/sql/cassandra.rst:204]: Similarly to debug_queries, this just checks for existence of the setting.

File: source/configuration_manual/sql/cassandra.rst [hunk 16]
```diff
@@ -0,0 +1,333 @@
...
+----------------
+- Default: 0
+- Values: :ref:`time`
+
+How long to idle before disconnecting.
+
+``connect_timeout``
+-------------------
+- Default: 5s
+- Values: :ref:`time`
```
Comment (sirainen) [inline source/configuration_manual/sql/cassandra.rst:216]: time_msecs (for all except idle_timeout and heartbeat_interval)

File: source/configuration_manual/sql/cassandra.rst [hunk 17]
```diff
@@ -0,0 +1,333 @@
...
+
+Connection timeout.
+
+``request_timeout``
+-------------------
+- Default: 60s
+- Values: :ref:`time`
```
Comment (sirainen) [inline source/configuration_manual/sql/cassandra.rst:223]: time_msecs

File: source/configuration_manual/sql/cassandra.rst [hunk 18]
```diff
@@ -0,0 +1,333 @@
...
+
+How long to wait for a query to finish.
+
+``warn_timeout``
+----------------
+- Default: 5s
+- Values: :ref:`time`
```
Comment (sirainen) [inline source/configuration_manual/sql/cassandra.rst:230]: time_msecs

File: source/configuration_manual/sql/cassandra.rst [hunk 19]
```diff
@@ -0,0 +1,333 @@
...
+----------------------
+- Default: 5s
+- Values: :ref:`time`
+
+How often to send keepalive packets to cassandra nodes.
+
+``execution_retry_interval``
+----------------------------
+- Default: 0
+- Values: :ref:`time`
```
Comment (sirainen) [inline source/configuration_manual/sql/cassandra.rst:244]: time_msecs

File: source/configuration_manual/sql/cassandra.rst [hunk 20]
```diff
@@ -0,0 +1,333 @@
...
+
+If the driver supports speculative execution policy, configures constant speculative execution policy.
+
+.. seealso:: https://docs.datastax.com/en/developer/java-driver/4.13/manual/core/speculative_execution/
+
+``execution_retry_times``
+-------------------------
+- Default: 0
+- Values: :ref:`time`
```
Comment (sirainen) [inline source/configuration_manual/sql/cassandra.rst:253]: uint

**Review by sirainen — CHANGES_REQUESTED** [review]

File: source/configuration_manual/sql/cassandra.rst (hunk 13 above)
Comment (cmouse) [reply source/configuration_manual/sql/cassandra.rst:117]: ?
```
                } else if (strcmp(key, "debug_queries") == 0) {
````

Comment (sirainen) [reply source/configuration_manual/sql/cassandra.rst]: debug_queries, not debuq_queries

File: source/settings/plugin/sql-sqlite.rst [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+=================
+
+Driver name is ``sqlite``.
+
+To compile support for this driver. you need sqlite library and headers.
+
+Supported options
+-----------------
+
+.. versionchanged:: v2.3.18
```
Comment (sirainen) [inline source/settings/plugin/sql-sqlite.rst:12]: I think this is v2.4 / v3.0

File: source/settings/plugin/sql-cassandra.rst [hunk 22]
```diff
@@ -0,0 +1,335 @@
...
+
+To compile support for this driver, you need to have DataStax C/C++ driver and headers installed.
+
+Supported options
+=================
+
+``host``
+--------
+- Default: <empty>
+- Values: :ref:`string`
```
Comment (sirainen) [inline source/settings/plugin/sql-cassandra.rst:19]: Should these be converted to the new settings formats?

**Review by sirainen — CHANGES_REQUESTED** [review]

File: source/settings/plugin/sql-cassandra.rst (hunk 22 above)
Comment (cmouse) [reply source/settings/plugin/sql-cassandra.rst:19]: Should.

File: source/settings/plugin/sql-mysql.rst [hunk 23]
```diff
@@ -0,0 +1,139 @@
+================
+SQL Driver MySQL
+================
+
+Driver name ``mysql``.
+
+Driver for MySQL / MariaDB server.
+
+To compile support for this driver, you need to have MySQL client library and headers installed.
+For MariaDB, you need to have compability headers installed.
```
Comment (jsoref) [inline source/settings/plugin/sql-mysql.rst:10]:
```suggestion
For MariaDB, you need to have compatibility headers installed.
```

**Comment by sirainen** [comment]
Replaced by https://github.com/dovecot/documentation/pull/374

**Pull request closed without merging**